### ISSUE TITLE
unable to use admin app since installing

### ISSUE THREAD

**Issue by greaneym** [prelegalwonder/openshift_web2py#6]
Hello,

since installing web2py, I am not able to access the admin tool.  When I access it, the error about it being disabled is displayed. I can't remember if I disabled it or it came that way.  So I am reading that the parameters_XXX.py file is what is needed and
I've put a hashed file in place in my git repo and pushed it and verified that it is there, but still the same error comes up.

When I do a git push, I believe that every service gets reset so I think the service restarted, and that means any cache should be removed?  Maybe there is some lingering cache file.

Can you verify, I should be able to run admin with this file in place.  What should the XXX number be?

Also, I am not able to get a ssh tunnel set up.  When I run the commands to set this up, the nohup.out contents show

ImportError: No module named gluon.widget
Traceback (most recent call last):
  File "web2py.py", line 16, in <module>
    import gluon.widget
ImportError: No module named gluon.widget

The other item I find is that I think the number list of open ports from openshift may not be correct, since some of these
numbers are listed and assigned in /etc/services on the openshift server.

thanks,
Margaret

**Comment by prelegalwonder** [comment]
Margaret,

Could you paste the contents of your wsgi/application file? It looks like there is an issue with gluon not being on your PYTHONPATH. 

Also, the parameters file should be parameters_443.py (whatever port you're using to access the application, in this case 443 since it's ssl to use admin).  Make sure you do a git add on the parameters file before committing so it pushes the file up there. You also should have a link from options_std.py -> options.py.

When this is setup correctly, you should not need to use an ssh tunnel. 

Regarding the number of ports open externally, it's only 443 and 80 that get routed to your gears. Internally, you have a limited range that you can bind to on $OPENSHIFT_INTERNAL_IP.

**Comment by greaneym** [comment]
Andrew, thanks for the info!

here are contents of wsgi/application:

#!/usr/bin/env python

import os
import sys

virtenv = os.environ['APPDIR'] + '/virtenv/'
os.environ['web2py_path'] = os.path.join(os.environ['OPENSHIFT_REPO_DIR'],
'wsgi', 'web2py')

sys.path.append(os.path.join(os.environ['OPENSHIFT_REPO_DIR'], 'wsgi',
'web2py'))
sys.path.append(os.path.join(os.environ['OPENSHIFT_REPO_DIR'], 'libs',
'gluon'))
sys.path.append(os.path.join(virtenv, 'lib/python2.6/site-packages'))

WEB2PY_LOG = os.path.join(os.environ['OPENSHIFT_LOG_DIR'],'web2py.log')
SQLITE_DIR = os.path.join(os.environ['OPENSHIFT_DATA_DIR'],'databases')
NEWRELIC_INI = os.path.join(os.environ['OPENSHIFT_REPO_DIR'], 'wsgi',
'web2py', 'newrelic.ini')

os.environ['PYTHON_EGG_CACHE'] = os.path.join(virtenv,
'lib/python2.6/site-packages')
virtualenv = os.path.join(virtenv, 'bin/activate_this.py')
try:
    execfile(virtualenv, dict(**file**=virtualenv))
except:
    pass

from gluon.settings import global_settings
import gluon.main

application = gluon.main.appfactory(wsgiapp=gluon.main.wsgibase,
                                    logfilename=None,
                                    profilerfilename=WEB2PY_LOG)

#NewRelic Monitoring
#import newrelic.agent
#newrelic.agent.initialize(NEWRELIC_INI)

#application = newrelic.agent.wsgi_application()(application)

**Comment by prelegalwonder** [comment]
Hm ok, I think the easiest way for me to assist with this is to get a snapshot of your gear. If you run: rhc snapshot save -a <yourapp> and upload the compressed tar file it creates to https://docs.google.com/folder/d/0ByP4lcjOqwkhSXNfVDMtOFdSOXM/edit  I'll import it and take a look.

**Comment by greaneym** [comment]
I think I would feel more comfortable to just try to reinstall this gear,
if you don't mind.

Do you know if I'm the only one having the problem with the admin access
using the default
python that openshift provides?

**Comment by prelegalwonder** [comment]
A few people have had issues with admin access but it came down to config and setup vs the provided runtime versions etc. I currently have a gear running with admin. 

Are you using the openshift_web2py template or are you moving anything around? If you moved gluon from libs/ back into web2py you'll need to tweak the application file so that it gets added to the path. 

I don't mind at all :)  I just figured being able to see the filesystem layout I'd be able to narrow in on issues quicker then going back and forth with things to check :)

Let me know how it works.

**Comment by greaneym** [comment]
Thanks Andrew,

I followed the directions on your README file and the openshift pages to
set up the app.
The only I changed was the parameters_xxx.py file, just trying different
number settings.  I
also added a directory containing an extra app besides the welcome app.
 The error I got
after installing when trying to connect admin was this,

admin disabled because unable to access password file

but assumed this is how it was supposed to work as first installed.

I am using a mac to connect, and will let you know how the delete app and
reinstall app goes.

Thank you for your help.

Margaret

**Comment by greaneym** [comment]
Andrew,

I have a new instance of web2y loaded and it comes with the 8080.py file
(with no hashed contents) and the admin gives the error
admin disabled because unable to access password file

I have not made any changes to the install.

Please verify that now for openshift, I should instead add a
parameters_443.py file, containing my hashed password  via change and
commit to my git repo, then push it and verify that it got there? and I
should remove the parameters_8080.py file?

If not this, what should I try next?

thanks

**Comment by prelegalwonder** [comment]
I would just create a symlink in the directory to the 8080 file like the following:

ln -s parameters_8080.py parameters_443.py
ln -s parameters_8080.py parameters_80.py
ln -s options_std.py options.py

Then git add . and try committing and pushing. This way you only have to modify the 8080 file when you want to change the hash.

**Comment by greaneym** [comment]
Andrew,

I am happy to report that with these changes, it worked fine.  I also made
the similar changes to the initial instance (I did not remove this but just
added a new instance of the web2py) and it works there also.

This is great!  Now I can keep on developing websites with web2py.  I
appreciate your help.

Margaret

**Comment by prelegalwonder** [comment]
Margaret, 

I'm glad to hear it's working for you :) Make sure you take note of the fact that changes made in the admin aren't persisted when you restart your gear so you'll want to export your app in the admin tool between doing so, or use the git support to push and pull from a git repo. 

I've been meaning to find a way to get the changes made via the Admin to commit directly to the gear git-repo but have not had time to look at this recently. Hopefully I will soon. I apologize for the inconvenience of having to use a workaround until that time.

Regards,
Andrew

**Comment by greaneym** [comment]
Andrew,

it must be just something about that directory, because I have another app
in the same directory but with a different name, including a sqlite dbase,
that never gets deleted.  I also do not understand why but will update you
if I figure
that out.

margaret

**Comment by greaneym** [comment]
Andrew,

It looks like people are using 2 methods for the DATA_DIR, either copying
or doing sym links to that dir.  Nam, the openshift tech who responds to
questions there posted this answer:
https://openshift.redhat.com/community/forums/openshift/one-newbie-question

I still don't understand yet what is going on there but this looks like it
would be the
answer.  The git repo has to be able to update a directory on the openshift
namespace,
and then that has to be in the /data in order to be persistent.  I am
looking to see if just a sym
link will work in google searches, but this link seems to be what Nam
proposes to work.

margaret

**Comment by prelegalwonder** [comment]
Thanks Margaret, I've updated the template to the latest release (2.0.2) and configured the Openshift action_hooks to save the applications directory in the $OPENSHIFT_DATA_DIR

Upon a git push it'll do an rsync to that directory, remove the local applications directory (in $OPENSHIFT_REPO_DIR), and then symlink to the one in the persistent ($OPENSHIFT_DATA_DIR) directory. 

Unfortunately I don't know how to get the changes made in the IDE to by pulled back down by a git pull which seems somewhat counter-productive to a PaaS. I'll continue to investigate as time permits. At least changes are saved across pushes now.

Regards,
Andrew

**Comment by greaneym** [comment]
Andrew, I will try this out. Thank you,

Margaret

**Comment by greaneym** [comment]
Andrew,

I made a new app on openshift and I could login with the admin app, but
when I clicked on the examples app, I got an error:

Traceback (most recent call last):
  File
"/var/lib/stickshift/8a1164fd28844e50b8804908fb9dcf2d/app-root/runtime/repo/libs/gluon/restricted.py",
line 209, in restricted
    exec ccode in environment
  File
"/var/lib/stickshift/8a1164fd28844e50b8804908fb9dcf2d/app-root/runtime/repo/wsgi/web2py/applications/admin/controllers/default.py",
line 1722, in <module>
  File
"/var/lib/stickshift/8a1164fd28844e50b8804908fb9dcf2d/app-root/runtime/repo/libs/gluon/globals.py",
line 185, in <lambda>
    self._caller = lambda f: f()
  File
"/var/lib/stickshift/8a1164fd28844e50b8804908fb9dcf2d/app-root/runtime/repo/wsgi/web2py/applications/admin/controllers/default.py",
line 949, in design
    in read_possible_languages(apath(app, r=request)).iteritems()
  File
"/var/lib/stickshift/8a1164fd28844e50b8804908fb9dcf2d/app-root/runtime/repo/libs/gluon/languages.py",
line 207, in read_possible_languages
    for filename in os.listdir(langdir):
OSError: [Errno 2] No such file or directory:
'/var/lib/stickshift/8a1164fd28844e50b8804908fb9dcf2d/app-root/runtime/repo/wsgi/web2py/applications/examples/languages'

I see on the web2py site that others are seeing this same error.

When I click on the welcome app, that does seem to load and I don't see an
error yet.

Margaret

**Comment by prelegalwonder** [comment]
Margaret,

In updating this template to the 2.0.2 version, I made some changes to have a symlink created to gluon in the libs directory during the build action_hook. 

You can either try updating to the latest version of the template or execute the following:

Edit <your-app>/.openshift/action_hooks/build

add the following line:

ln -s $OPENSHIFT_REPO_DIR/libs/gluon $OPENSHIFT_REPO_DIR/wsgi/web2py/

If you're not using a unix based os, be careful to ensure that the ln tool is properly creating the symlink. Alec ran into some issues with the cygwin version of ln that was behaving oddly. 

Regards,
Andrew

**Comment by AlecTaylor** [comment]
To be exact, on Windows delete the `parameters_8080.py` then run `web2py.py -p 8080 -a [yourpasswordhere]`.

Confirm that a new `parameters_8080.py` is in place before committing -> pushing. It should then all work fine for you =)

**Comment by greaneym** [comment]
Andrew,

I'm having some network problems with my provider today, but have noticed
that since I
reinstalled with the openshift build since your last e-mail, the app failed
and I was not able to see it's http url.  When I logged in there wasn't any
repo.

Not sure yet what caused this. I looked at your script which does copy app
contents into data, but not sure if entire repo needs to also be in /data.

I have not been able to do a successful git push today and have tried
several times.

But I am going to wait until I have a good network connection to continue.

Margaret
